### PR TITLE
DecimalUtil which is a Decimal version of NumberUtil, and extend Decimal functionality

### PR DESCRIPTION
**Pull request by ShawTim**

### CODE REVIEW

File: packages/tempus-core-services/src/datastructures/Decimal.ts [hunk 1]
```diff
@@ -76,16 +104,40 @@ export default class Decimal {
     const [integral, fraction] = str.split('.');
 
     const lastDigit = fraction.charAt(fractionDigits);
-    if (Number(lastDigit) < 5) {
-      return this.toTruncated(fractionDigits);
+
+    if (this.gte(0)) {
+      // positive number
```
Comment (danieledamiani) [inline packages/tempus-core-services/src/datastructures/Decimal.ts:109]: `gte` is non-negative

File: packages/tempus-core-services/src/datastructures/Decimal.ts [hunk 2]
```diff
@@ -76,16 +104,40 @@ export default class Decimal {
...
-        const str = utils.formatUnits(bigNum.add(1), fractionDigits);
-        const [outputIntegral, outputFraction] = str.split('.');
-        return `${outputIntegral}.${outputFraction.padEnd(fractionDigits, '0')}`;
+      // negative number
+      if (Number(lastDigit) <= 5) {
+        // no need to be rounded
+        return this.toTruncated(fractionDigits);
       } else {
-        return BigNumber.from(integral).add(1).toString();
+        if (fractionDigits > 0) {
```
Comment (danieledamiani) [inline packages/tempus-core-services/src/datastructures/Decimal.ts:131]: how do we reach this branch?

**Review by danieledamiani — APPROVED** [review]

File: packages/tempus-core-services/src/utils/DecimalUtils.ts [hunk 3]
```diff
@@ -0,0 +1,51 @@
+import Decimal, { Numberish } from '../datastructures/Decimal';
+import { reverseString } from './reverseString';
+
+const MULTIPLIER_LOOKUP = [
```
Comment (danieledamiani) [inline packages/tempus-core-services/src/utils/DecimalUtils.ts:4]: maybe a map would be more efficient

File: packages/tempus-core-services/src/utils/DecimalUtils.ts [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+  { value: '1000000', symbol: 'M' },
+  { value: '1000', symbol: 'k' },
+];
+
+export default class DecimalUtils {
+  // round for multiplier (e.g. 9876 -> 9.88k), trauncate for fraction (e.g. 9.876 -> 9.87)
+  static formatWithMultiplier(value: Numberish, fractionDigits: number = 0): string {
+    const decimal = new Decimal(value);
+    const decimalAbs = decimal.abs();
+    const multiplier = MULTIPLIER_LOOKUP.find(item => decimalAbs.gte(item.value));
```
Comment (danieledamiani) [inline packages/tempus-core-services/src/utils/DecimalUtils.ts:18]: if we use a map we can do
```
const decimalAbsLength = decimal.toString().length;
const multiplier = MULTIPLIER_LOOKUP[decimalAbsLength];
```

File: packages/tempus-core-services/src/datastructures/Decimal.ts (hunk 1 above)
Comment (ShawTim) [reply packages/tempus-core-services/src/datastructures/Decimal.ts]: updated comment

File: packages/tempus-core-services/src/utils/DecimalUtils.ts (hunk 3 above)
Comment (ShawTim) [reply packages/tempus-core-services/src/utils/DecimalUtils.ts:4]: updated as suggested

File: packages/tempus-core-services/src/utils/DecimalUtils.ts (hunk 4 above)
Comment (ShawTim) [reply packages/tempus-core-services/src/utils/DecimalUtils.ts]: we still need `MULTIPLIER_LOOKUP.find()` but updated as suggested to check the string length instead

**Review by lazarmitic — APPROVED** [review]

**Pull request merged**